test/Test/Pos/Block/Logic/VarSpec.hs

```diff
@@ -265,24 +271,130 @@ blockPropertyScenarioGen m = do
-        pretty e
-    BlockScenarioDbChanged d ->
+        pretty e <> "\n" <>
+        pretty bsec
+    BlockScenarioDbChanged bsec d ->
         let DbNotEquivalentToSnapshot snapId dbDiff = d in
         stopProperty $
             "Block scenario resulted in a change to the blockchain" <>
             " relative to the " <> show snapId <> " snapshot:\n" <>
-            show dbDiff
+            show dbDiff <> "\n" <>
+            pretty bsec
+
+----------------------------------------------------------------------------
+-- Multi-epoch
+----------------------------------------------------------------------------
+
+-- Input: the amount of blocks after crossing.
+applyThroughEpochSpec :: Int -> Spec
+applyThroughEpochSpec afterCross = do
+    prop applyThroughEpochDesc (applyThroughEpochProp afterCross)
+  where
+    applyThroughEpochDesc =
+      "apply a sequence of blocks that spans through epochs (additional blocks after crossing: " ++
+      show afterCross ++ ")"
+
+applyThroughEpochProp :: Int -> BlockProperty ()
+applyThroughEpochProp afterCross = do
+    scenario <- blockPropertyScenarioGen $ do
+        let
+            approachEpochEdge =
+                pathSequence mempty . OldestFirst . NE.fromList $
+                replicate (fromIntegral epochSlots - 1) "a"
+            crossEpochEdge =
+                pathSequence (NE.last $ getOldestFirst approachEpochEdge) $
+                OldestFirst . NE.fromList $
+                -- 2 blocks to ensure that we cross,
+                -- then some additional blocks
+                replicate (afterCross + 2) "x"
+        emitBlockApply BlockApplySuccess approachEpochEdge
+        emitBlockApply BlockApplySuccess crossEpochEdge
+    runBlockScenarioAndVerify scenario
+
+----------------------------------------------------------------------------
+-- Forks
+----------------------------------------------------------------------------
+
+singleForkSpec :: ForkDepth -> Spec
+singleForkSpec fd = do
+    prop singleForkDesc (singleForkProp fd)
+  where
+    singleForkDesc =
+      "a blockchain of length q<=(9.5*k) blocks can switch to a fork " <>
+      "of length j>i with a common prefix i, rollback depth d=q-i"
+
+singleForkProp :: ForkDepth -> BlockProperty ()
+singleForkProp fd = do
+    scenario <- blockPropertyScenarioGen $ genSingleFork fd
+    runBlockScenarioAndVerify scenario
+
+data ForkDepth = ForkShort | ForkMedium | ForkDeep
+
+genSingleFork :: forall g m. (RandomGen g, Monad m) => ForkDepth -> BlockEventGenT g m ()
+genSingleFork fd = do
+    let k = fromIntegral blkSecurityParam :: Int
+    -- 'd' is how deeply in the chain the fork starts. In other words, it's how many
+    -- blocks we're going to rollback (therefore must be >1).
+    d <- getRandomR $ case fd of
+        ForkShort  -> (1, if k > 1 then k-1 else 1)
+        ForkMedium -> (if k > 2 then k - 2 else 1, k+2)
+        ForkDeep   -> (k+1, div (k*3) 2 + 1)
+    -- the depth must be <=k for a successful rollback.
```

Comment:
it must be `<=k` for a *successful* rollback, we're testing failing ones as well.